safedelete/queryset.py

```diff
@@ -44,10 +127,27 @@ def undelete(self, force_policy=None):
             :py:func:`safedelete.models.SafeDeleteModel.undelete`
         """
         assert self.query.can_filter(), "Cannot use 'limit' or 'offset' with undelete."
-        # TODO: Replace this by bulk update if we can (need to call pre/post-save signal)
-        for obj in self.all():
-            obj.undelete(force_policy=force_policy)
-        self._result_cache = None
+
+        all_objs = self.all()
+        all_objs_id_list = list(all_objs.values_list('id', flat=True))
+        current_policy = force_policy or all_objs.model._safedelete_policy
+
+        with transaction.atomic():
+            if utils.is_safedelete_cls(all_objs.model):
+                if current_policy == SOFT_DELETE_CASCADE:
+                    soft_undelete_cascade_policy_action(all_objs, send_signals)
+
+                # TODO: WTF is going on? if do not print(all_objs.query), the test_undelete_queryset() will fail.
+                #   couldn't fix by myself, waiting for an answer with big the exciting
```

Comment:
I don't understand the full PR, but if I had  to guess, print is forcing a refresh of the object cache. Instead of printing, you could try `refresh_from_db()`.

https://docs.djangoproject.com/en/4.0/ref/models/instances/#django.db.models.Model.refresh_from_db

Reply:
Thank you, for ur answer. The QuerySet object has no method 'refresh_from_db()', different sources, I have checked, recommend to use 'all()' method, but it is already used in undelete/delete as:
`all_objs = self.all()`
where 'self' - is instance of SafeDeleteQueryset class

Reply:
undelete (and also delete) is done without filtering without printing query:

```
UPDATE "safedelete_softdeletemodel" SET "deleted" = NULL
```
That's probably because `query._filter_visibility` is called when printing the query. Adding `self.query._filter_visibility()` after the assert fix the issue.